### PR TITLE
1746/feat/new-identification-types

### PR DESCRIPTION
**Pull request by Noam6421**
### What in this PR? 
* New Identification Types options for IdentificationType field.

* New route: GET /investigationInfo/identificationTypes

* Edit DB functions:
1.  update_contact_person
2.  update_contact_persons
3.  update_static_info

* DB Changes: 
1. new column on Identification_type table: id (type number) && new fields of the new Identification Types.
2. new FK Identification_type on tables: person and investigated_patient.

* Main UI changes:
1. Static info of investigation now supports the new identification types.
2. InteractionsTab: dropdown list of the identification types instead of toggle.
3. ContactQuestioningTab: new identification types added.
4. Contacts Excel: has hidden id column of the identification types.

### CODE REVIEW

File: client/src/Utils/Contacts/ContactIdValidationSchema.ts [hunk 1]
```diff
@@ -1,26 +1,34 @@
...
+                is: IdentificationTypesCodes.PALESTINE_ID,
+                then: yup.string()
+                    .nullable()
+                    .test('isValid', invalidPalestineIdText, (id) => isPalestineIdValid(id)),
+                otherwise:
+                    yup.string()
+                    .nullable()
+                    .test('isValid', invalidOtherIdText, (id) => isOtherIdValid(id)),
+            })
+        })
```
Comment (y-golde) [inline client/src/Utils/Contacts/ContactIdValidationSchema.ts:31]: maybe It's more effective to use when as a function - see examples in [here](https://github.com/jquense/yup#mixedwhenkeys-string--arraystring-builder-object--value-schema-schema-schema)

File: client/src/Utils/Contacts/useContactFields.ts [hunk 2]
```diff
@@ -55,14 +58,23 @@ const useContactFields = (contactStatus?: InteractedContact['contactStatus']) =>
...
+        if (contactIdType === identificationTypes[1].id && !isPassportValid(contactIdNumber)) {
             return { valid: false, error: 'שדה דרכון אינו תקין' };
         };
-
+        if ((contactIdType === identificationTypes[2].id || contactIdType === identificationTypes[3].id ) && !isOtherIdValid(contactIdNumber)) {
+            return { valid: false, error: 'שדה מזהה אינו תקין' };
+        };
+        if (contactIdType === identificationTypes[4].id && !isPalestineIdValid(contactIdNumber)) {
+            return { valid: false, error: 'שדה ת"ז פלסטינית אינו תקין' };
+        };
```
Comment (y-golde) [inline client/src/Utils/Contacts/useContactFields.ts:76]: consider using a switch statement , or seperate into a function like `validateId(idNum , idType)` with something like an enum to determine the eventual error

File: client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/ContactSection/Schemas/contactQuestioningPersonal.ts [hunk 3]
```diff
@@ -7,7 +7,6 @@ import ContactIdValidationSchema from 'Utils/Contacts/ContactIdValidationSchema'
 import { PHONE_NUMBER_REGEX, NOT_REQUIRED_PHONE_NUMBER_REGEX} from 'commons/Regex/Regex';
 
 export const contactQuestioningPersonal = {
-    [InteractedContactFields.IDENTIFICATION_TYPE]: yup.string(),
```
Comment (y-golde) [inline client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/ContactSection/Schemas/contactQuestioningPersonal.ts:10]: why did you remove it?

File: client/src/components/App/Content/InvestigationForm/useInvestigationForm.ts [hunk 4]
```diff
@@ -90,7 +91,20 @@ const useInvestigationForm = (): useInvestigationFormOutcome => {
...
                 contactTypesLogger.error(`got errors in server result: ${error}`, Severity.HIGH);
             });
-    }
+    };
+
+    const fetchIdentificationTypes = () => {
+        const identificationTypesLogger = logger.setup('Fetching Identification Types');
+        identificationTypesLogger.info('launching request to get identification types', Severity.LOW);
+        axios.get('/investigationInfo/identificationTypes')
+            .then((result: any) => {
```
Comment (y-golde) [inline client/src/components/App/Content/InvestigationForm/useInvestigationForm.ts:100]: why any?

File: client/src/components/App/Content/InvestigationForm/useInvestigationForm.ts [hunk 5]
```diff
@@ -158,6 +172,10 @@ const useInvestigationForm = (): useInvestigationFormOutcome => {
         });
     };
 
+    useEffect(() => {
+        fetchIdentificationTypes();
+    }, []);
```
Comment (y-golde) [inline client/src/components/App/Content/InvestigationForm/useInvestigationForm.ts:177]: maybe add reference to if the redux store? or use Thunk? just check if the current store state is empty and then if it is - fetch it

File: client/src/models/enums/IdentificationTypesCodes.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+enum IdentificationTypesCodes {
+    ID=1,
+    PASSPORT=2,
+    OTHER=3,
+    MOSSAD=4,
+    PALESTINE_ID=5
+};
```
Comment (y-golde) [inline client/src/models/enums/IdentificationTypesCodes.ts:7]: If we're getting it from the server why keep a local record of all of the types? think about what happens if you want to add another ID type.  
I don't ask you to change it just to rethink about the system's structure :\

File: server/src/DBService/InvestigationInfo/Mutation.ts [hunk 7]
```diff
@@ -73,7 +73,7 @@ mutation deleteinvestigationReasonId ($epidemiologyNumberInput: Int!, $oldComple
 `;
 
 export const UPDATE_INVESTIGATION_STATIC_INFO = gql`
-mutation updateInvestigationStaticInfo ($fullNameInput: String, $identificationTypeInput: String, $identityNumberInput: String, $epidemiologyNumberInput: Int!) {
+mutation updateInvestigationStaticInfo ($fullNameInput: String, $identificationTypeInput: Int, $identityNumberInput: String, $epidemiologyNumberInput: Int!) {
```
Comment (y-golde) [inline server/src/DBService/InvestigationInfo/Mutation.ts:76]: can it be null? if it can't write `Int!`

File: server/src/ScriptRunner/Scripts/index/BLK_new_identification_types.sql [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+-- Updating values in identification_type table
+
+UPDATE public.identification_type
+	SET id=1
+	WHERE type='ת"ז';
+	
+UPDATE public.identification_type
+	SET id=2
+	WHERE type='דרכון';
+
```
Comment (y-golde) [inline server/src/ScriptRunner/Scripts/index/BLK_new_identification_types.sql:15]: thank godddddd 👏

File: server/src/ScriptRunner/Scripts/index/BLK_new_identification_types.sql [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+	
+-- Add constraint of FK in investigated_patient table
+
+ALTER TABLE public.investigated_patient
+    ADD CONSTRAINT identity_type_fk FOREIGN KEY (identity_type) REFERENCES public.identification_type (id);
+	
+	
+-- Delete duplicate function of update_contact_person
+
+DROP FUNCTION public.update_contact_person(json, integer);
```
Comment (y-golde) [inline server/src/ScriptRunner/Scripts/index/BLK_new_identification_types.sql:111]: IF EXISTS

File: server/src/ScriptRunner/Scripts/index/BLK_new_identification_types.sql [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+-- Updating identity_type column type
+
+ALTER TABLE public.investigated_patient
+	ALTER COLUMN identity_type TYPE integer 
+	USING identity_type::integer;
+	
+-- Add constraint of FK in investigated_patient table
+
+ALTER TABLE public.investigated_patient
+    ADD CONSTRAINT identity_type_fk FOREIGN KEY (identity_type) REFERENCES public.identification_type (id);
```
Comment (y-golde) [inline server/src/ScriptRunner/Scripts/index/BLK_new_identification_types.sql:106]: IF NOT EXISTS

File: server/src/ScriptRunner/Scripts/index/BLK_new_identification_types.sql [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+-- Updating identification_type column type
+
+ALTER TABLE public.person
+	ALTER COLUMN identification_type TYPE integer 
+	USING identification_type::integer;
+
+-- Add constraint of FK in person table
+
+ALTER TABLE public.person
+    ADD CONSTRAINT identification_type_fk FOREIGN KEY (identification_type) REFERENCES public.identification_type (id);
```
Comment (y-golde) [inline server/src/ScriptRunner/Scripts/index/BLK_new_identification_types.sql:71]: IF NOT EXISTS

Comment (y-golde) [reply server/src/ScriptRunner/Scripts/index/BLK_new_identification_types.sql:71]: https://stackoverflow.com/questions/6801919/postgres-add-constraint-if-it-doesnt-already-exist

Comment (y-golde) [reply server/src/ScriptRunner/Scripts/index/BLK_new_identification_types.sql:71]: see the answer here

**Review by y-golde — CHANGES_REQUESTED** [review]
Great work overall 👏

File: client/src/Utils/Contacts/ContactIdValidationSchema.ts (hunk 1 above)
Comment (Noam6421) [reply client/src/Utils/Contacts/ContactIdValidationSchema.ts:31]: I'm using when the problem is that I have to compare to various values so more than one when is needed.

File: client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/ContactSection/Schemas/contactQuestioningPersonal.ts (hunk 3 above)
Comment (Noam6421) [reply client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/ContactSection/Schemas/contactQuestioningPersonal.ts:10]: because this validation isn't necessary anymore, there validation in other places in the code

File: client/src/components/App/Content/InvestigationForm/useInvestigationForm.ts (hunk 4 above)
Comment (Noam6421) [reply client/src/components/App/Content/InvestigationForm/useInvestigationForm.ts:100]: we always use any type for result because its the server respond and there are a lot of fields there so doing type to that it's hard

File: client/src/components/App/Content/InvestigationForm/useInvestigationForm.ts (hunk 5 above)
Comment (Noam6421) [reply client/src/components/App/Content/InvestigationForm/useInvestigationForm.ts:177]: added check if the current store state is empty and then just if only empty then fetch it

File: client/src/models/enums/IdentificationTypesCodes.ts (hunk 6 above)
Comment (Noam6421) [reply client/src/models/enums/IdentificationTypesCodes.ts:7]: I completely agree with you, the reason we added enum is that we needed access to the identification codes in a file that don't allow use of selector

File: server/src/DBService/InvestigationInfo/Mutation.ts (hunk 7 above)
Comment (Noam6421) [reply server/src/DBService/InvestigationInfo/Mutation.ts:76]: yes it can be null

File: client/src/Utils/Contacts/ContactIdValidationSchema.ts (hunk 1 above)
Comment (y-golde) [reply client/src/Utils/Contacts/ContactIdValidationSchema.ts:31]: yea but you can use a function as the second handler - look at the link that I refrenced

File: client/src/components/App/Content/InvestigationForm/useInvestigationForm.ts (hunk 4 above)
Comment (y-golde) [reply client/src/components/App/Content/InvestigationForm/useInvestigationForm.ts:100]: yea but doesn't identificationTypes supposed to return the Store's type?

File: client/src/models/enums/IdentificationTypesCodes.ts (hunk 6 above)
Comment (y-golde) [reply client/src/models/enums/IdentificationTypesCodes.ts:7]: 🤷

File: server/src/ScriptRunner/Scripts/index/BLK_new_identification_types.sql (hunk 11 above)
Comment (Noam6421) [reply server/src/ScriptRunner/Scripts/index/BLK_new_identification_types.sql:71]: yes now used it, thank you!

File: client/src/Utils/Contacts/ContactIdValidationSchema.ts (hunk 1 above)
Comment (Noam6421) [reply client/src/Utils/Contacts/ContactIdValidationSchema.ts]: changed it according to the example

**Review by y-golde — APPROVED** [review]

**Review by barnahum8 — APPROVED** [review]
